scp.1
.\"
.\" scp.1
.\"
.\" Author: Tatu Ylonen <ylo@cs.hut.fi>
.\"
.\" Copyright (c) 1995 Tatu Ylonen <ylo@cs.hut.fi>, Espoo, Finland
.\"                    All rights reserved
.\"
.\" Created: Sun May  7 00:14:37 1995 ylo
.\"
.\" $OpenBSD: scp.1,v 1.116 2026/09/03 16:02:45 job Exp $
.\"
.Dd $Mdocdate: September 3 2026 $
.Dt SCP 1
.Os
.Sh NAME
.Nm scp
.Nd OpenSSH secure file copy
.Sh SYNOPSIS
.Nm scp
.Op Fl 346ABCOpqRrTv
.Op Fl c Ar cipher
.Op Fl D Ar sftp_server_path
.Op Fl F Ar ssh_config
.Op Fl i Ar identity_file
.Op Fl J Ar destination
.Op Fl l Ar limit
.Op Fl o Ar ssh_option
.Op Fl P Ar port
.Op Fl S Ar program
.Op Fl X Ar sftp_option
.Ar source ... target
.Sh DESCRIPTION
.Nm
copies files between hosts on a network.
.Pp
.Nm
uses the SFTP protocol over an
.Xr ssh 1
connection for data transfer, and uses the same authentication and provides
the same security as a login session.
.Pp
.Nm
will ask for passwords or passphrases if they are needed for
authentication.
.Pp
The
.Ar source
and
.Ar target
may be specified as a local pathname, a remote host with optional path
in the form
.Sm off
.Oo user @ Oc host : Op path ,
.Sm on
or a URI in the form
.Sm off
.No scp:// Oo user @ Oc host Oo : port Oc Op / path .
.Sm on
Local file names can be made explicit using absolute or relative pathnames
to avoid
.Nm
treating file names containing
.Sq :\&
as host specifiers.
.Pp
When copying between two remote hosts, if the URI format is used, a
.Ar port
cannot be specified on the
.Ar target
if the
.Fl R
option is used.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl 3
Copies between two remote hosts are transferred through the local host.
This mode is the default,
but see also the
.Fl R
option for copying data directly between two remote hosts.
Note that when using the legacy SCP protocol (via the
.Fl O
flag), this option
selects batch mode for the second host as
.Nm
cannot ask for passwords or passphrases for both hosts.
.It Fl 4
Forces
.Nm
to use IPv4 addresses only.
.It Fl 6
Forces
.Nm
to use IPv6 addresses only.
.It Fl A
Allows forwarding of
.Xr ssh-agent 1
to the remote system.
The default is not to forward an authentication agent.
.It Fl B
Selects batch mode (prevents asking for passwords or passphrases).
.It Fl C
Compression enable.
Passes the
.Fl C
flag to
.Xr ssh 1
to enable compression.
.It Fl c Ar cipher
Selects the cipher to use for encrypting the data transfer.
This option is directly passed to
.Xr ssh 1 .
.It Fl D Ar sftp_server_path
Connect directly to a local SFTP server program rather than a
remote one via
.Xr ssh 1 .
This option may be useful in debugging the client and server.
.It Fl F Ar ssh_config
Specifies an alternative
per-user configuration file for
.Nm ssh .
This option is directly passed to
.Xr ssh 1 .
.It Fl i Ar identity_file
Selects the file from which the identity (private key) for public key
authentication is read.
This option is directly passed to
.Xr ssh 1 .
.It Fl J Ar destination
Connect to the target host by first making an
.Nm
connection to the jump host described by
.Ar destination
and then establishing a TCP forwarding to the ultimate destination from
there.
Multiple jump hops may be specified separated by comma characters.
This is a shortcut to specify a
.Cm ProxyJump
configuration directive.
This option is directly passed to
.Xr ssh 1 .
.It Fl l Ar limit
Limits the used bandwidth, specified in Kbit/s.
.It Fl O
Use the legacy SCP protocol for file transfers instead of the SFTP protocol.
Forcing the use of the SCP protocol may be necessary for servers that do
not implement SFTP, for backwards-compatibility for particular filename
wildcard patterns and for expanding paths with a
.Sq ~
prefix for older SFTP servers.
.It Fl o Ar ssh_option
Can be used to pass options to
.Nm ssh
in the format used in
.Xr ssh_config 5 .
This is useful for specifying options
for which there is no separate
.Nm scp
command-line flag.
For full details of the options listed below, and their possible values, see
.Xr ssh_config 5 .
.Pp
.Bl -tag -width Ds -offset indent -compact
.It AddKeysToAgent
.It AddressFamily
.It BatchMode
.It BindAddress
.It BindInterface
.It CASignatureAlgorithms
.It CanonicalDomains
.It CanonicalizeFallbackLocal
.It CanonicalizeHostname
.It CanonicalizeMaxDots
.It CanonicalizePermittedCNAMEs
.It CertificateFile
.It ChannelTimeout
.It CheckHostIP
.It Ciphers
.It ClearAllForwardings
.It Compression
.It ConnectTimeout
.It ConnectionAttempts
.It ControlMaster
.It ControlPath
.It ControlPersist
.It DynamicForward
.It EnableEscapeCommandline
.It EnableSSHKeysign
.It EscapeChar
.It ExitOnForwardFailure
.It FingerprintHash
.It ForkAfterAuthentication
.It ForwardAgent
.It ForwardX11
.It ForwardX11Timeout
.It ForwardX11Trusted
.It GSSAPIAuthentication
.It GSSAPIDelegateCredentials
.It GatewayPorts
.It GlobalKnownHostsFile
.It HashKnownHosts
.It Host
.It HostKeyAlgorithms
.It HostKeyAlias
.It HostbasedAcceptedAlgorithms
.It HostbasedAuthentication
.It Hostname
.It IPQoS
.It IdentitiesOnly
.It IdentityAgent
.It IdentityFile
.It IgnoreUnknown
.It Include
.It KbdInteractiveAuthentication
.It KbdInteractiveDevices
.It KexAlgorithms
.It KnownHostsCommand
.It LocalCommand
.It LocalForward
.It LogLevel
.It LogVerbose
.It MACs
.It NoHostAuthenticationForLocalhost
.It NumberOfPasswordPrompts
.It ObscureKeystrokeTiming
.It PKCS11Provider
.It PasswordAuthentication
.It PermitLocalCommand
.It PermitRemoteOpen
.It Port
.It PreferredAuthentications
.It ProxyCommand
.It ProxyJump
.It ProxyUseFdpass
.It PubkeyAcceptedAlgorithms
.It PubkeyAuthentication
.It RekeyLimit
.It RemoteCommand
.It RemoteForward
.It RequestTTY
.It RequiredRSASize
.It RevokedHostKeys
.It SecurityKeyProvider
.It SendEnv
.It ServerAliveCountMax
.It ServerAliveInterval
.It SessionType
.It SetEnv
.It StdinNull
.It StreamLocalBindMask
.It StreamLocalBindUnlink
.It StrictHostKeyChecking
.It SyslogFacility
.It TCPKeepAlive
.It Tag
.It Tunnel
.It TunnelDevice
.It UpdateHostKeys
.It User
.It UserKnownHostsFile
.It VerifyHostKeyDNS
.It VisualHostKey
.It XAuthLocation
.El
.It Fl P Ar port
Specifies the port to connect to on the remote host.
Note that this option is written with a capital
.Sq P ,
because
.Fl p
is already reserved for preserving the times and mode bits of the file.
.It Fl p
Preserves modification times, access times, and file mode bits from the
source file.
.It Fl q
Quiet mode: disables the progress meter as well as warning and diagnostic
messages from
.Xr ssh 1 .
.It Fl R
Copies between two remote hosts are transferred through the local host
by default.
This option instead copies between two remote hosts by connecting to the origin
host and executing
.Nm
there.
This requires that
.Nm
running on the origin host can authenticate to the destination host without
requiring a password.
.It Fl r
Recursively copy entire directories.
Note that
.Nm
follows symbolic links encountered in the tree traversal.
.It Fl S Ar program
Name of
.Ar program
to use for the encrypted connection.
The program must understand
.Xr ssh 1
options.
.It Fl T
Disable strict filename checking.
By default when copying files from a remote host to a local directory
.Nm
checks that the received filenames match those requested on the command-line
to prevent the remote end from sending unexpected or unwanted files.
Because of differences in how various operating systems and shells interpret
filename wildcards, these checks may cause wanted files to be rejected.
This option disables these checks at the expense of fully trusting that
the server will not send unexpected filenames.
.It Fl v
Verbose mode.
Causes
.Nm
and
.Xr ssh 1
to print debugging messages about their progress.
This is helpful in
debugging connection, authentication, and configuration problems.
.It Fl X Ar sftp_option
Specify an option that controls aspects of SFTP protocol behaviour.
The valid options are:
.Bl -tag -width Ds
.It Cm nrequests Ns = Ns Ar value
Controls how many concurrent SFTP read or write requests may be in progress
at any point in time during a download or upload.
By default 64 requests may be active concurrently.
.It Cm buffer Ns = Ns Ar value
Controls the maximum buffer size for a single SFTP read/write operation used
during download or upload.
By default a 32KB buffer is used.
.El
.El
.Pp
Usage of SCP protocol can be blocked by creating a world-readable
.Ar /etc/ssh/disable_scp
file. If this file exists, when SCP protocol is in use (either remotely or 
via the
.Fl O
option), the program will exit.
.Sh EXIT STATUS
.Ex -std scp
.Sh SEE ALSO
.Xr sftp 1 ,
.Xr ssh 1 ,
.Xr ssh-add 1 ,
.Xr ssh-agent 1 ,
.Xr ssh-keygen 1 ,
.Xr ssh_config 5 ,
.Xr sftp-server 8 ,
.Xr sshd 8
.Sh HISTORY
.Nm
is based on the rcp program in
.Bx
source code from the Regents of the University of California.
.Pp
Since OpenSSH 9.0,
.Nm
has used the SFTP protocol for transfers by default.
.Sh AUTHORS
.An Timo Rinne Aq Mt tri@iki.fi
.An Tatu Ylonen Aq Mt ylo@cs.hut.fi
.Sh CAVEATS
The legacy SCP protocol (selected by the
.Fl O
flag) requires execution of the remote user's shell to perform
.Xr glob 3
pattern matching.
This requires careful quoting of any characters that have special meaning to
the remote shell, such as quote characters.


ChangeLog
commit fdda125f5349558c17ef572f2afced1980b7d208
Author: Damien Miller <djm@mindrot.org>
Date:   Tue Oct 6 19:01:18 2026 +1100

    .depend

commit ce43722bcc9f5426c6469ae06527b809a9cb8920
Author: Damien Miller <djm@mindrot.org>
Date:   Tue Oct 6 18:57:28 2026 +1100

    update RPM spec files

commit b27b24bd295594816adcc979788eca001140532c
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Mon Oct 5 09:54:05 2026 +0000

    upstream: openssh-10.6
    
    OpenBSD-Commit-ID: 907104c6a6e058860522bf752663d71b37bfa43f

commit 1058092dba9145325a5f50793570fc754b375c3d
Author: dtucker@openbsd.org <dtucker@openbsd.org>
Date:   Sat Oct 3 05:08:49 2026 +0000

    upstream: Add test for proxycommand percent expansions.
    
    OpenBSD-Regress-ID: be34cb5cf08f1025413df248cb7a65e18ca3e2a1

commit 630144f410985ce9bad03929556b3c364bc0e255
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Wed Sep 23 21:45:51 2026 +0000

    upstream: backout regress bits related to multiplexing change
    
    OpenBSD-Regress-ID: 3fdaea9d640ce1d4950223b66f26f6f1ab46f4c9

commit ac6ad1afd1c3531fad2e96a00c84bd9867cf5561
Author: Damien Miller <djm@mindrot.org>
Date:   Tue Oct 6 16:48:54 2026 +1100

    remove the --disable-fd-passing configure option
    
    Also add a deprecation warning when the lack FD passing support and also
    requires root for the post-auth sshd-session process

commit b37c095aa39bb2c01754673dec50a684a536db24
Author: Damien Miller <djm@mindrot.org>
Date:   Tue Oct 6 16:43:39 2026 +1100

    disable features when post-auth sshd runs as root
    
    On platforms that require root privilege for the post-authentication
    sshd-session process (e.g. OpenServer 5, QNX 6), disable and restrict
    a number of features that assume user privilege.
    
    This includes GatewayPorts, StreamLocalForwarding and -R forwardings
    binding to ports <1024

commit 4b29da5952b1ede28f0d225cf722dee4923ba04c
Author: Darren Tucker <dtucker@dtucker.net>
Date:   Mon Oct 5 17:08:15 2026 +1100

    Don't automatically enable FORTIFY_SOURCE.
    
    It can cause problems on some platforms, in particular NetBSD <11 since
    it will cause function pointer comparisons in atomicio to fail and some
    things including scp to hang.  Previously we had a workaround but that
    was removed in a765b86d.  See NetBSD bug 45200 and pkgsrc bug pkg/60563.

commit 0c4c9a7b320bd7ff61c4a2dc345cb2e27e9663b2
Author: dtucker@openbsd.org <dtucker@openbsd.org>
Date:   Mon Oct 5 06:03:40 2026 +0000

    upstream: Further restrict the characters allowed in a command-line
    
    supplied user name, disallowing '$' and '\'. Reported by SecBuddyF KeenLab
    Tencent (CodeBuddy Security).
    
    OpenBSD-Commit-ID: 13671c178f492af63b5c1296f284818c7809ed9a

commit 15289cc38956aa74d4947174a68c10ac9ad2157f
Author: Darren Tucker <dtucker@dtucker.net>
Date:   Sat Oct 3 17:20:09 2026 +1000

    Remove NetBSD 9.0 test target.
    
    It can no longer install the sudo package and is covered by the
    selfhosted VMs.

commit 4f4bee9ec673b29949ec0e79c92cd56b61f772e6
Author: Darren Tucker <dtucker@dtucker.net>
Date:   Sat Oct 3 17:12:13 2026 +1000

    Replace native ARM runner with remote runner.
    
    Native ARM32 runners are no longer supported by Github.

commit e0f28caca5a77e84bef02c9f6476d2d4bd31b937
Author: Darren Tucker <dtucker@dtucker.net>
Date:   Sat Oct 3 16:18:30 2026 +1000

    Pull older NetBSD sudo package from archive.

commit d1d37e27a14f3336e36666d59c4b3e1c76cce328
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Sat Oct 3 00:46:58 2026 +0000

    upstream: make StreamLocalBindMask properly first-match-wins;
    
    spotted while fixing bz4013
    
    OpenBSD-Commit-ID: 35ef4524da0d03a439751f7bcd4847f76d93ec7f

commit bc00e06e0cda509c06b47ee4d1551efc904894a1
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Sat Oct 3 00:46:29 2026 +0000

    upstream: make StreamLocalBindMask properly respect Host/Match
    
    blocks and make it first-match-wins as documented. bz4013
    
    OpenBSD-Commit-ID: e2efc85b42bbf7067ece4f1ab12d7d02ec6c3e12

commit 87f0cd1892e501f835dd210abea5461807c8deba
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Thu Oct 1 07:10:56 2026 +0000

    upstream: start process of deprecating the -R flag. This was the
    
    old way of performing a remote-to-remote copy that was basically executed scp
    on the remote host. It barely worked (needing agent forwarding enabled or
    usable credentials on the remote host) and has largely been replaced by a
    better SFTP-protocol remote-to-remote copy that runs through the host
    performing the copy.
    
    We'll disable this option in a release or two; ok dtucker@
    
    OpenBSD-Commit-ID: dc273300651e581c3db18edf21f2b05ceac60fd7

commit 88fe0b074ee0c0cdf6d87f8b86b4c959e7d2bdbc
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Thu Oct 1 07:08:05 2026 +0000

    upstream: mention default KDF rounds is now 32
    
    OpenBSD-Commit-ID: 0d17bf0ad02c8c0d897d9d01d1e4eb0f65ea749c

commit f938c78d490cb7556aa8ca1625e9b9bd1f963ed3
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Thu Oct 1 07:07:49 2026 +0000

    upstream: Implement a maximum number of KDF rounds that will be
    
    accepted when writing an OpenSSH-format private key or when loading one. This
    limit is set pretty high (1<<20), but ensures that a service that is passed a
    bad key with an ridiculously high number of rounds will _eventually_ complete
    parsing it.
    
    Also bump the default number of KDF rounds from 24 to 32 (this is a
    linear increase, not like bcrypt(3) which is exponential).
    
    Pointed out by Aris Adamantiadis
    
    OpenBSD-Commit-ID: 843ff37b066b2879f4579a784e42a3c9d22b2ddc

commit 39b9bd1cb7d32428842c67e7e8b217027c16962b
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Thu Oct 1 04:17:39 2026 +0000

    upstream: fix the bit length of ML-DSA 44/Ed25519 keys that was
    

regress/scp.sh
#	$OpenBSD: scp.sh,v 1.19 2023/09/08 05:50:57 djm Exp $
#	Placed in the Public Domain.

tid="scp"

#set -x

COPY2=${OBJ}/copy2
DIR=${COPY}.dd
DIR2=${COPY}.dd2
COPY3=${OBJ}/copy.glob[123]
DIR3=${COPY}.dd.glob[456]
DIFFOPT="-rN"

# Figure out if diff does not understand "-N"
if ! diff -N ${SRC}/scp.sh ${SRC}/scp.sh 2>/dev/null; then
	DIFFOPT="-r"
fi

maybe_add_scp_path_to_sshd

SRC=`dirname ${SCRIPT}`
cp ${SRC}/scp-ssh-wrapper.sh ${OBJ}/scp-ssh-wrapper.scp
chmod 755 ${OBJ}/scp-ssh-wrapper.scp
export SCP # used in scp-ssh-wrapper.scp

scpclean() {
	rm -rf ${COPY} ${COPY2} ${DIR} ${DIR2} ${COPY3} ${DIR3}
	mkdir ${DIR} ${DIR2} ${DIR3}
	chmod 755 ${DIR} ${DIR2} ${DIR3}
}

# Create directory structure for recursive copy tests.
forest() {
	scpclean
	rm -rf ${DIR2}
	cp ${DATA} ${DIR}/copy
	ln -s ${DIR}/copy ${DIR}/copy-sym
	mkdir ${DIR}/subdir
	cp ${DATA} ${DIR}/subdir/copy
	ln -s ${DIR}/subdir ${DIR}/subdir-sym
}

for mode in scp sftp ; do
	tag="$tid: $mode mode"
	if test $mode = scp ; then
		scpopts="-O -S ${OBJ}/scp-ssh-wrapper.scp"
	else
		scpopts="-qs -D ${SFTPSERVER}"
	fi

	verbose "$tag: simple copy local file to local file"
	scpclean
	$SCP $scpopts ${DATA} ${COPY} || fail "copy failed"
	cmp ${DATA} ${COPY} || fail "corrupted copy"

	verbose "$tag: simple copy local file to remote file"
	scpclean
	$SCP $scpopts ${DATA} somehost:${COPY} || fail "copy failed"
	cmp ${DATA} ${COPY} || fail "corrupted copy"

	verbose "$tag: simple copy remote file to local file"
	scpclean
	$SCP $scpopts somehost:${DATA} ${COPY} || fail "copy failed"
	cmp ${DATA} ${COPY} || fail "corrupted copy"

	verbose "$tag: copy local file to remote file in place"
	scpclean
	cp ${DATA} ${COPY}
	$SCP $scpopts ${COPY} somehost:${COPY} || fail "copy failed"
	cmp ${DATA} ${COPY} || fail "corrupted copy"

	verbose "$tag: copy remote file to local file in place"
	scpclean
	cp ${DATA} ${COPY}
	$SCP $scpopts somehost:${COPY} ${COPY} || fail "copy failed"
	cmp ${DATA} ${COPY} || fail "corrupted copy"

	verbose "$tag: copy local file to remote file clobber"
	scpclean
	cat ${DATA} ${DATA} > ${COPY}
	$SCP $scpopts ${DATA} somehost:${COPY} || fail "copy failed"
	ls -l $DATA $COPY
	cmp ${DATA} ${COPY} || fail "corrupted copy"

	verbose "$tag: copy remote file to local file clobber"
	scpclean
	cat ${DATA} ${DATA} > ${COPY}
	$SCP $scpopts somehost:${DATA} ${COPY} || fail "copy failed"
	cmp ${DATA} ${COPY} || fail "corrupted copy"

	verbose "$tag: simple copy local file to remote dir"
	scpclean
	cp ${DATA} ${COPY}
	$SCP $scpopts ${COPY} somehost:${DIR} || fail "copy failed"
	cmp ${COPY} ${DIR}/copy || fail "corrupted copy"

	verbose "$tag: simple copy local file to local dir"
	scpclean
	cp ${DATA} ${COPY}
	$SCP $scpopts ${COPY} ${DIR} || fail "copy failed"
	cmp ${COPY} ${DIR}/copy || fail "corrupted copy"

	verbose "$tag: simple copy remote file to local dir"
	scpclean
	cp ${DATA} ${COPY}
	$SCP $scpopts somehost:${COPY} ${DIR} || fail "copy failed"
	cmp ${COPY} ${DIR}/copy || fail "corrupted copy"

	verbose "$tag: recursive local dir to remote dir"
	forest
	$SCP $scpopts -r ${DIR} somehost:${DIR2} || fail "copy failed"
	diff ${DIFFOPT} ${DIR} ${DIR2} || fail "corrupted copy"

	verbose "$tag: recursive local dir to local dir"
	forest
	rm -rf ${DIR2}
	cp ${DATA} ${DIR}/copy
	$SCP $scpopts -r ${DIR} ${DIR2} || fail "copy failed"
	diff ${DIFFOPT} ${DIR} ${DIR2} || fail "corrupted copy"

	verbose "$tag: recursive remote dir to local dir"
	forest
	rm -rf ${DIR2}
	cp ${DATA} ${DIR}/copy
	$SCP $scpopts -r somehost:${DIR} ${DIR2} || fail "copy failed"
	diff ${DIFFOPT} ${DIR} ${DIR2} || fail "corrupted copy"

	# we don't want to run this test if openssl is not installed
	# if isn't then the usage text won't have the "vZ" in the
	# output
	if $SCP -Z 2>&1 | grep "vZ" > /dev/null 2>&1
	then
	    verbose "$tag: resume remote dir to local dir"
	    scpclean
	    rm -rf ${DIR2}
	    cp ${DATA} ${DIR}/copy1
	    cp ${DATA} ${DIR}/copy2
	    cp ${DATA} ${DIR}/copy3
	    $SCP -r $scpopts somehost:${DIR} ${DIR2} || fail "copy failed"
	    truncate --size=-512 ${DIR2}/copy1
	    truncate --size=+512 ${DIR2}/copy2
	    $SCP -Z $scpopts somehost:${DIR}/* ${DIR2} || fail "resume failed"
	    for i in $(cd ${DIR} && echo *); do
		cmp ${DIR}/$i ${DIR2}/$i || fail "corrupted resume copy"
	    done
	fi

	if $SCP -Z 2>&1 | grep "vZ" > /dev/null 2>&1
	then
	    verbose "$tag: resume local dir to remote dir"
	    scpclean
	    rm -rf ${DIR2}
	    cp ${DATA} ${DIR}/copy1
	    cp ${DATA} ${DIR}/copy2
	    cp ${DATA} ${DIR}/copy3
	    $SCP -r $scpopts ${DIR} somehost:${DIR2} || fail "copy failed"
	    truncate --size=-512 ${DIR2}/copy1
	    truncate --size=+512 ${DIR2}/copy2
	    $SCP -Z $scpopts ${DIR}/* somehost:${DIR2} || fail "resume failed"
	    for i in $(cd ${DIR} && echo *); do
		cmp ${DIR}/$i ${DIR2}/$i || fail "corrupted resume copy"
	    done
	fi

	verbose "$tag: unmatched glob file local->remote"
	scpclean
	$SCP $scpopts ${DATA} somehost:${COPY3} || fail "copy failed"
	cmp ${DATA} ${COPY3} || fail "corrupted copy"

	verbose "$tag: unmatched glob file remote->local"
	# NB. no clean
	$SCP $scpopts somehost:${COPY3} ${COPY2} || fail "copy failed"
	cmp ${DATA} ${COPY2} || fail "corrupted copy"

	verbose "$tag: unmatched glob dir recursive local->remote"
	scpclean
	rm -rf ${DIR3}
	cp ${DATA} ${DIR}/copy
	cp ${DATA} ${DIR}/copy.glob[1234]
	$SCP $scpopts -r ${DIR} somehost:${DIR3} || fail "copy failed"
	diff ${DIFFOPT} ${DIR} ${DIR3} || fail "corrupted copy"

	verbose "$tag: unmatched glob dir recursive remote->local"
	# NB. no clean
	rm -rf ${DIR2}
	$SCP $scpopts -r somehost:${DIR3} ${DIR2} || fail "copy failed"
	diff ${DIFFOPT} ${DIR} ${DIR2} || fail "corrupted copy"

	verbose "$tag: shell metacharacters"
	scpclean
	(cd ${DIR} && \
	 touch '`touch metachartest`' && \
	 $SCP $scpopts *metachar* ${DIR2} 2>/dev/null; \
	 [ ! -f metachartest ] ) || fail "shell metacharacters"

	if [ ! -z "$SUDO" ]; then
		verbose "$tag: skipped file after scp -p with failed chown+utimes"
		scpclean
		cp -p ${DATA} ${DIR}/copy
		cp -p ${DATA} ${DIR}/copy2
		cp ${DATA} ${DIR2}/copy
		chmod 660 ${DIR2}/copy
		$SUDO chown root ${DIR2}/copy
		$SCP -p $scpopts somehost:${DIR}/\* ${DIR2} >/dev/null 2>&1
		$SUDO diff ${DIFFOPT} ${DIR} ${DIR2} || fail "corrupted copy"
		$SUDO rm ${DIR2}/copy
	fi

	for i in 0 1 2 3 4 5 6 7; do
		verbose "$tag: disallow bad server #$i"
		SCPTESTMODE=badserver_$i
		export DIR SCPTESTMODE
		scpclean
		$SCP $scpopts somehost:${DATA} ${DIR} >/dev/null 2>/dev/null
		[ -d {$DIR}/rootpathdir ] && fail "allows dir relative to root dir"
		[ -d ${DIR}/dotpathdir ] && fail "allows dir creation in non-recursive mode"

		scpclean
		$SCP -r $scpopts somehost:${DATA} ${DIR2} >/dev/null 2>/dev/null
		[ -d ${DIR}/dotpathdir ] && fail "allows dir creation outside of subdir"

		scpclean
		$SCP -pr $scpopts somehost:${DATA} ${DIR2} >/dev/null 2>/dev/null
		[ ! -w ${DIR2} ] && fail "allows target root attribute change"

		scpclean
		$SCP $scpopts somehost:${DATA} ${DIR2} >/dev/null 2>/dev/null
		[ -e ${DIR2}/extrafile ] && fail "allows unauth object creation"
		rm -f ${DIR2}/extrafile
	done

	verbose "$tag: detect non-directory target"
	scpclean
	echo a > ${COPY}
	echo b > ${COPY2}
	$SCP $scpopts ${DATA} ${COPY} ${COPY2}
	cmp ${COPY} ${COPY2} >/dev/null && fail "corrupt target"
done

scpclean
rm -f ${OBJ}/scp-ssh-wrapper.scp
